runtime/tests/utils/utils.go

```diff
@@ -50,13 +50,12 @@ const ImportedLocation = common.StringLocation("imported")
 //
 // If the objects are not equal, this function prints a human-readable diff.
 func AssertEqualWithDiff(t *testing.T, expected, actual any) {
-	if !assert.Equal(t, expected, actual) {
-		// the maximum levels of a struct to recurse into
-		// this prevents infinite recursion from circular references
-		deep.MaxDepth = 100
-
-		diff := deep.Equal(expected, actual)
+	// the maximum levels of a struct to recurse into
+	// this prevents infinite recursion from circular references
+	deep.MaxDepth = 100
+	diff := deep.Equal(expected, actual)
 
+	if !assert.Nil(t, diff) {
```

Comment:
It's unclear to me why exactly these changes exercised this, but it's possible for `assert.Equal` to consider unequal two objects that `deep.Equal` says are equivalent. This results in nonsensical errors where a test fails, but the "diff" between the two objects is empty.

Reply:
That's odd. Does the PR itself exercise this if the change here is not made? If not, do you have a way to reproduce it in some other way?

Reply:
It turns out that `deep.Equal` does not report all differences (IDK why). I opened #2819 to replace it.